### DISCUSSION TITLE
DDF output for data augmentation transforms

### DISCUSSION THREAD

**Discussion by mariakawula** [Project-MONAI/MONAI#3933] [Q&A]
Hi all,

is there a way to output the dense displacement field that was used to deform images for data augmentation? I defined a Compose transform that contains, among others, RandAffine and BSpline based deformation field (RandomElasticDeformation from torchio).

I can extract the affine matrix that in principle could be used to get the ddf but I don't have anything for B-splines.

**Comment by wyli** [comment]
not sure if I understand your question correctly, but you may need some modification of the `__call__` to store the ddf while doing the transform, an example is here https://github.com/Project-MONAI/MONAI/pull/1793, more specifically https://github.com/Project-MONAI/MONAI/blob/42111622ad9e9b674ed0ac09abbe83ac2c97088c/monai/transforms/spatial/dictionary.py#L844-L856 cc @rijobro

**Comment by ebrahimebrahim** [comment]
For creating a DDF out of affine transforms specifically, the implementation of AffineHead, which is part of GlobalNet, could be helpful. See here:
https://github.com/Project-MONAI/MONAI/blob/f15ff935ba9f9de47ec1a398ccf413336fa2e0ab/monai/networks/nets/regunet.py#L267-L278

**Comment by mariakawula** [comment]
Thank you all for your suggestions. I will try to implement them and let you know if it worked out

**Comment by mariakawula** [comment]
That would be the transform

```
from monai.transforms import RandAffined, AddChanneld, CenterSpatialCropd
from torchio.transforms import RandomElasticDeformation, Compose

train_transform = Compose(
        [
            LoadImaged(keys=["image", "label"], reader="itkreader"),
            AddChanneld(keys=["image", "label"]),

            RandAffined(keys=["image", "label"], prob=args.aug_prob, rotate_range=np.deg2rad(args.rotate_range),
                        padding_mode="zeros", translate_range=args.translate_range, scale_range=args.scale_factor),
            RandomElasticDeformation(include=["image", "label"], num_control_points=args.n_control_points,
                                     max_displacement=args.max_displacement, p=args.aug_prob),
            CenterSpatialCropd(keys=["image", "label"], roi_size=args.vol_dims),
            ToTensord(keys=["image", "label"])
        ]
    )
```
